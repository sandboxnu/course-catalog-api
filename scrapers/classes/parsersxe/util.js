import $ from "cheerio";
import _ from "lodash";
import Request from "../../request";
import macros from "../../../utils/macros";

const request = new Request("util");

function validCell(el) {
  return el.type === "tag" && ["th", "td"].includes(el.name);
}

/**
 * Modify a string to avoid collisions with set
 * @param {[String]} set array to avoid collisions with
 * @param {String} value String to uniquify
 * appends a number to end of the string such that it doesn't collide
 */
function uniquify(set, value) {
  if (set.includes(value)) {
    let append = 1;
    while (set.includes(value + append)) {
      append++;
    }
    return value + append;
  }
  return value;
}

/**
 * Parse a table using it's head (or first row) as keys
 * @param {Cheerio} table Cheerio object of table
 * @returns A list of {key: value} where key comes from header
 */
function parseTable(table) {
  if (table.length !== 1 || table[0].name !== "table") {
    return [];
  }

  //includes both header rows and body rows
  const rows = $("tr", table).get();
  if (rows.length === 0) {
    macros.error("zero rows???");
    return [];
  }

  //the headers
  const heads = rows[0].children.filter(validCell).reduce((acc, element) => {
    const head = $(element).text().trim().toLowerCase().replace(/\s/gi, "");
    const uniqueHead = uniquify(acc, head);
    acc.push(uniqueHead);
    return acc;
  }, []);

  //add the other rows
  const ret = [];

  rows.slice(1).forEach((row) => {
    const values = row.children.filter(validCell).map((el) => {
      return $(el).text();
    });
    if (values.length >= heads.length) {
      // TODO look into which classes trigger this
      // macros.log('warning, table row is longer than head, ignoring some content');
    }

    ret.push(_.zipObject(heads, values));
  });
  return ret;
}

async function getCookiesForSearch(termCode) {
  // first, get the cookies
  // https://jennydaman.gitlab.io/nubanned/dark.html#studentregistrationssb-clickcontinue-post
  const clickContinue = await request.post({
    url: "https://nubanner.neu.edu/StudentRegistrationSsb/ssb/term/search?mode=search",
    form: {
      term: termCode,
    },
    cache: false,
  });

  if (clickContinue.body.regAllowed === false) {
    macros.error(
      `failed to get cookies (from clickContinue) for the term ${termCode}`,
      clickContinue
    );
  }

  const cookiejar = request.jar();
  for (const cookie of clickContinue.headers["set-cookie"]) {
    cookiejar.setCookie(
      cookie,
      "https://nubanner.neu.edu/StudentRegistrationSsb/"
    );
  }
  return cookiejar;
}

export default {
  parseTable: parseTable,
  getCookiesForSearch: getCookiesForSearch,
};
