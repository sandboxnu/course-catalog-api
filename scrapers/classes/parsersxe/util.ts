import $ from "cheerio";
import _ from "lodash";
import Request from "../../request";
import macros from "../../../utils/macros";
import { CookieJar } from "tough-cookie";

const requestObj = new Request("util");

function validCell(el: cheerio.Element): boolean {
  return el.type === "tag" && ["th", "td"].includes(el.name);
}

/**
 * Modify a string to avoid collisions with set
 * @param {[String]} set array to avoid collisions with
 * @param {String} value String to uniquify
 * appends a number to end of the string such that it doesn't collide
 */
function uniquify(set: string[], value: string): string {
  if (set.includes(value)) {
    let append = 1;
    while (set.includes(value + append)) {
      append++;
    }
    return value + append;
  }
  return value;
}

/**
 * Parse a table using it's head (or first row) as keys
 * @param {Cheerio} table Cheerio object of table
 * @returns A list of {key: value} where key comes from header
 */
function parseTable(table: cheerio.Cheerio): Record<string, string>[] {
  // Empty table
  if (table.length !== 1) {
    return [];
  }
  // Non-table
  if (!("name" in table[0]) || table[0].name !== "table") {
    return [];
  }

  //includes both header rows and body rows
  const rows: cheerio.TagElement[] = $("tr", table).get();
  if (rows.length === 0) {
    macros.error("zero rows???");
    return [];
  }

  //the headers
  const heads: string[] = rows[0].children
    .filter(validCell)
    .reduce((acc: string[], element) => {
      const head: string = $(element)
        .text()
        .trim()
        .toLowerCase()
        .replace(/\s/gi, "");
      const uniqueHead = uniquify(acc, head);
      acc.push(uniqueHead);
      return acc;
    }, []);

  // add the other rows
  const ret: Record<string, string>[] = [];

  rows.slice(1).forEach((row: cheerio.TagElement) => {
    const values: string[] = row.children
      .filter(validCell)
      .map((el) => $(el).text());
    if (values.length > heads.length) {
      // TODO look into which classes trigger this
      macros.warn(
        "Table row is longer than head, ignoring some content",
        heads,
        values
      );
    }

    ret.push(_.zipObject(heads, values) as Record<string, string>);
  });

  return ret;
}

async function getCookiesForSearch(termId: string): Promise<CookieJar> {
  const cookieJar = new CookieJar();
  // first, get the cookies
  // https://jennydaman.gitlab.io/nubanned/dark.html#studentregistrationssb-clickcontinue-post
  const clickContinue = await requestObj.post(
    "https://nubanner.neu.edu/StudentRegistrationSsb/ssb/term/search?mode=search",
    {
      form: {
        term: termId,
      },
      cacheRequests: false,
      cookieJar: cookieJar,
    }
  );

  const bodyObj = JSON.parse(clickContinue.body);

  if (bodyObj.regAllowed === false) {
    macros.error(
      `failed to get cookies (from clickContinue) for the term ${termId}`,
      clickContinue.body
    );
  }

  for (const cookie of clickContinue.headers["set-cookie"]) {
    cookieJar.setCookie(
      cookie,
      "https://nubanner.neu.edu/StudentRegistrationSsb/"
    );
  }
  return cookieJar;
}

export default {
  parseTable: parseTable,
  getCookiesForSearch: getCookiesForSearch,
  uniquify: uniquify,
};
