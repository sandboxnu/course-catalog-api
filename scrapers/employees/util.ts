import URI from "urijs";
import macros from "../../utils/macros";

// http://stackoverflow.com/questions/4009756/how-to-count-string-occurrence-in-string/7924240#7924240
export function occurrences(
  string: string,
  subString: string,
  allowOverlapping: boolean
): number {
  string += "";
  subString += "";
  if (subString.length <= 0) {
    return string.length + 1;
  }

  let n = 0;
  let pos = 0;
  const step = allowOverlapping ? 1 : subString.length;

  while (pos >= 0) {
    pos = string.indexOf(subString, pos);
    if (pos >= 0) {
      ++n;
      pos += step;
    } else {
      break;
    }
  }
  return n;
}

export function parseNameWithSpaces(name: string): {
  firstName: string;
  lastName: string;
} {
  // Standardize spaces.
  name = name.trim().replace(/\s+/gi, " ");

  // Generate first name and last name
  const spaceCount = occurrences(name, " ", false);
  const splitName = name.split(" ");

  if (spaceCount === 0) {
    macros.warn("0 spaces found in name", name);
    return null;
  }

  // Handles firstName, lastName and firstName, middleName, lastName

  if (spaceCount > 2) {
    macros.log(
      `${name} has more than 1 space in their name. Using first and last word.`
    );
  }

  return {
    firstName: splitName[0],
    lastName: splitName[splitName.length - 1],
  };
}

// Standardizes email addresses found across different pages
// Removes a 'mailto:' from the beginning
// Ensures the email contains a @
export function standardizeEmail(email: string): null | string {
  if (!email) {
    return null;
  }

  if (email.startsWith("mailto:")) {
    email = email.slice("mailto:".length);
  }

  if (!email.includes("@") || email.includes(" ")) {
    return null;
  }

  if (email.endsWith("@neu.edu")) {
    email = `${email.split("@")[0]}@northeastern.edu`;
  }

  return email.toLowerCase().trim();
}

export function standardizePhone(phone: string): null | string {
  if (!phone) {
    return null;
  }

  phone = phone.trim();

  if (phone.startsWith("tel:")) {
    phone = phone.slice("tel:".length).trim();
  }

  let digitsOnly = phone.replace(/[^0-9]/gi, "");

  if (phone.startsWith("+1") && digitsOnly.length === 11) {
    digitsOnly = digitsOnly.slice(1);
  }

  if (digitsOnly.length !== 10) {
    return null;
  }

  return digitsOnly;
}

// Parses the google scholar id from a link that should contain a google scholar link.
// Get the Google Scholar ID with this: https://scholar.google.com/citations?user=[id here]
export function parseGoogleScholarLink(link: string): null | string {
  if (!link) {
    return null;
  }

  const userId = new URI(link).query(true).user;
  if (!userId && link) {
    macros.log("Error parsing google url", link);
    return null;
  }
  return userId;
}

// Gets the base hostname from a url.
// fafjl.google.com -> google.com
// subdomain.bob.co -> bob.co
// bob.co -> bob.co
// This could be improved by using public lists of top-level domains.
export function getBaseHost(url: string): null | string {
  const homepage = new URI(url).hostname();
  if (!homepage || homepage === "") {
    macros.error("could not find homepage of", url);
    return null;
  }

  const match = homepage.match(/[^.]+\.[^.]+$/i);
  if (!match) {
    macros.error("homepage match failed...", homepage);
    return null;
  }
  return match[0];
}
