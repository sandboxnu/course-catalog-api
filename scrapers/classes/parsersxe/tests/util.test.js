import fs from "fs-extra";
import path from "path";
import cheerio from "cheerio";
import util from "../util";

describe("parseTable", () => {
  it("pulls out right data", async () => {
    const body = await fs.readFile(
      path.join(__dirname, "data", "util", "1.html"),
      "utf8",
    );

    const rawTable = cheerio.load(body)("table");
    const parsedTable = util.parseTable(rawTable);
    expect(parsedTable).toMatchSnapshot();
  });

  it("ignores columns too wide and blank cells", async () => {
    const body = await fs.readFile(
      path.join(__dirname, "data", "util", "2.html"),
      "utf8",
    );

    const rawTable = cheerio.load(body)("table");
    const parsedTable = util.parseTable(rawTable);
    expect(parsedTable).toMatchSnapshot();
  });

  it("uniquifies the head", async () => {
    const body = await fs.readFile(
      path.join(__dirname, "data", "util", "3.html"),
      "utf8",
    );

    const rawTable = cheerio.load(body)("table");
    const parsedTable = util.parseTable(rawTable);
    expect(parsedTable).toMatchSnapshot();
  });
});

it("uniquifies", () => {
  expect(util.uniquify(["1", "11"], "1")).toBe("12");
});

describe("parseTable", () => {
  it("table without name/table", () => {
    expect(util.parseTable([])).toEqual([]);
    expect(
      util.parseTable([
        {
          name: "not table",
        },
      ]),
    ).toEqual([]);
  });

  it("table with no rows", () => {
    expect(
      util.parseTable([
        {
          name: "table",
          rows: [],
        },
      ]),
    ).toEqual([]);
  });
});
