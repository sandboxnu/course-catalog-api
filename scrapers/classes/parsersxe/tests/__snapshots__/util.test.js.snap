// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTable ignores columns too wide and blank cells 1`] = `
[
  {
    "days": undefined,
    "time": "11:00 am - 11:50 am",
    "type": "Class",
  },
  {
    "days": "MW",
    "time": "",
    "type": "Class",
  },
]
`;

exports[`parseTable pulls out right data 1`] = `
[
  {
    "days": "MWF",
    "time": "11:00 am - 11:50 am",
    "type": "Class",
  },
  {
    "days": "MW",
    "time": "12:00 pm - 1:50 pm",
    "type": "Class",
  },
]
`;

exports[`parseTable uniquifies the head 1`] = `
[
  {
    "days": undefined,
    "time": "Class",
    "time1": "11:00 am - 11:50 am",
  },
  {
    "days": "MW",
    "time": "Class",
    "time1": "",
  },
]
`;
